### PR TITLE
Make compatible with Python 3

### PR DESCRIPTION
**Pull request by lucaswiman**
@svituz: Similar to @eheinz's work in #19. This PR adds compatibility with Python 3, without adding any external dependencies like six.

I tried to make the commits pretty atomic and homogeneous. The only difficult change was the encoding changes in `mllp.py`, which I'll comment on below.

## Testing

I verified that the tests pass on python 2.7, 3.4 and 3.5 on OS X. I updated the travis.yml file so it should also run on those versions.

### CODE REVIEW

File: hl7apy/mllp.py [hunk 1]
```diff
@@ -47,50 +54,54 @@ def __str__(self):
...
     def setup(self):
-        self.sb = "\x0b"
-        self.eb = "\x1c"
-        self.cr = "\x0d"
-        self.validator = re.compile(self.sb + "(([^\r]+\r)*([^\r]+\r?))" + self.eb + self.cr)
+        self.sb = b"\x0b"
+        self.eb = b"\x1c"
+        self.cr = b"\x0d"
+        self.validator = re.compile(
+            self.sb.decode('ascii') + r"(([^\r]+\r)*([^\r]+\r?))" + self.eb.decode('ascii') + self.cr.decode('ascii'))
```
Comment (lucaswiman) [inline hl7apy/mllp.py]: Note that this validator can only run against decoded text values (unicode in Python 2.7), so we need to decode the sequence identifiers to build a `str` instead of `bytes`.

File: hl7apy/mllp.py [hunk 2]
```diff
@@ -47,50 +54,54 @@ def __str__(self):
...
             self.request.close()
             return
 
-        if line[0] != self.sb:  # First MLLP char
+        if io.getvalue()[:1] != self.sb:  # First MLLP char
             self.request.close()
             return
 
-        while line[-2:] != end_seq:
+        while io.getvalue()[-2:] != end_seq:
```
Comment (lucaswiman) [inline hl7apy/mllp.py]: I think this should be safe in both LATIN-1 (because each byte is also a character) and UTF-8 (since the bytes in `end_seq` are in the 0-127 range).

File: hl7apy/mllp.py [hunk 3]
```diff
@@ -47,50 +54,54 @@ def __str__(self):
...
+        io = BytesIO()
         try:
-            line = self.request.recv(3)
+            io.write(self.request.recv(3))
         except socket.timeout:
             self.request.close()
             return
 
-        if line[0] != self.sb:  # First MLLP char
+        if io.getvalue()[:1] != self.sb:  # First MLLP char
```
Comment (lucaswiman) [inline hl7apy/mllp.py]: Note the need for a slice there since `b'foo'[1]` in python 3 is equivalent to `chr('foo'[1])` in Python 2.

File: hl7apy/mllp.py [hunk 4]
```diff
@@ -47,50 +54,54 @@ def __str__(self):
 
 
 class _MLLPRequestHandler(StreamRequestHandler):
+    encoding = 'utf-8'
```
Comment (lucaswiman) [inline hl7apy/mllp.py]: UTF-8 seems to be the most common encoding in the US. From what I've read, LATIN-1 is the most common for HL7 in Europe. I don't have an immediate need for this module, so I'm happy to set it to whatever encoding you'd prefer.

`ascii` would be the most restrictive, and would work on the common subset of UTF-8 and LATIN-1, so might be a good compromise.

File: hl7apy/mllp.py (hunk 1 above)
Comment (svituz) [inline hl7apy/mllp.py]: Here we can use format or join which is better.
`"".join([self.sb.decode('ascii'), r"(([^\r]+\r)*([^\r]+\r?))", self.eb.decode('ascii'), self.cr.decode('ascii')])`

File: hl7apy/mllp.py (hunk 4 above)
Comment (svituz) [inline hl7apy/mllp.py]: Ok for encoding in utf-8. We plan to let the user specify the encoding he/she wants to use but for now we can use utf-8

**Comment by svituz** [comment]
Hello Lucas,
thank you for your good work. Python 3 support is one of the main goal we wanted to accomplish.
I commented some lines that I would like you to change before merging your pull request. Also I noticed that you continue your work in your branch so if you want you can update your pull request with new changes.
Thanks again,
Vittorio

File: examples/lab62/server.py [hunk 5]
```diff
@@ -19,9 +19,16 @@
...
+from __future__ import absolute_import
+from __future__ import print_function
+
 import re
 
+try:
+    import SocketServer as socketserver
+except ImportError:
+    import socketserver
+
```
Comment (lucaswiman) [inline examples/lab62/server.py]: Updated the import. 

> BTW: this example should be changed to use the mllp module. It is legacy code written before we implemented MLLP

I don't really know enough about MLLP to rework the example, since we only use the parser. Reworking the example seems unrelated to updating to Python 3. Is this a requirement to get this merged?

**Comment by lucaswiman** [comment]
I've addressed the code review comments, other than updating the server.py file to use MLLP, which I don't feel very qualified to do.

Please let me know if there are additional changes you'd like me to do.

**Comment by svituz** [comment]
Hello Lucas,
I merged your pull requests. Regarding the server.py file, that was just a comment about the fact that we need to change the example, I didn't mean you had do that.

**Comment by lucaswiman** [comment]
Thanks very much @svituz!
